### DISCUSSION TITLE
Add jemalloc to the linux image

### DISCUSSION THREAD

**Discussion by hassila** [SwiftPackageIndex/SwiftPackageIndex-Server#2184] [Ideas]
As suggested in:
https://swiftpackageindex.com/SwiftPackageIndex/SPIManifest/0.13.0/documentation/spimanifest/commonusecases#linuxImages

I'd like to ask if 'jemalloc' could be added as an addition to the image? 

We use this for the Benchmark package for measuring memory allocations and any package that uses ordo-one/Benchmark will fail to build (even though our CI works fine, as we have it as a dependency there).

**Comment by finestructure** [comment]
Hi @hassila thanks for raising this! Do you have the ubuntu (`apt-get`) package name to hand? We're using `focal` as the base image. That'd be very helpful and ensure we've got the right thing in place.

I see

```
libjemalloc-dev - development files and documentation for jemalloc
libjemalloc1 - general-purpose scalable concurrent malloc(3) implementation
```

and I suspect we'd need `libjemalloc-dev`, right?

**Reply by finestructure** [reply]
This is for https://swiftpackageindex.com/ordo-one/package-benchmark, right? Just so I can leave a note when adding it.

**Reply by hassila** [reply]
Hi,

Yes, we use this on Ubuntu:

`sudo apt-get install -y libjemalloc-dev`

And, yes it's for `ordo-one/package-benchmark`, that's correct (and for any other package using it as an dependency).

Thanks!

**Reply by hassila** [reply]
(macOS is a different question - don't know how you build there or if it's possible to do anything about it - jemalloc installed with homebrew works great though but unfortunately we need Ventura as we use swift 5.7 runtime features)

**Reply by finestructure** [reply]
Yea, since our Linux builds are docker image based an update to the image immediately reflects in current and future builders without further effort. Plus we have the option of supporting multiple environments for package builds to run in.

On macOS these dependencies are global and we'd have to make sure we apply them to the current and future runners, that's why we've held off adding any OS-level dependencies for now.

**Reply by finestructure** [reply]
Ok, that's fixed the `main` build:

<img width="525" alt="CleanShot 2022-12-12 at 14 06 21@2x" src="https://user-images.githubusercontent.com/65520/207052541-9edd2c93-316e-4373-a470-72f27a2492ec.png">

Since we only have a handful of additional dependencies right now and they're not conflicting in any way, I've made the `spi-image` our base compile image (to be deployed). That means the `image: registry.gitlab.com/finestructure/spi-images:basic-5.7-latest` directive won't be required anymore and the 0.5.0 should turn green as well, once we've deployed this and triggered a rebuild.

**Reply by finestructure** [reply]
Ok, that's switched over to the new base image now:

<img width="525" alt="CleanShot 2022-12-12 at 16 36 26@2x" src="https://user-images.githubusercontent.com/65520/207086974-0ebf725d-49ba-4289-bd40-35d9c1416373.png">

**Comment by hassila** [comment]
Awesome, thanks! Then I can drop the .spi.yml (except for DocC, but there it's blocked by DocC being dependent on 5.6 building on macOS AFAIU?) 

I tried to push it to use 5.7/linux:

```
version: 1
builder:
  configs:
  - platform: linux
    swift_version: '5.7'
    image: registry.gitlab.com/finestructure/spi-images:basic-5.7-latest
  - documentation_targets: [BenchmarkSupport]
    platform: linux
```

but no go as it seems.

**Reply by hassila** [reply]
Oh, sorry - I just saw 'We currently only support generating documentation on macOS and iOS' now, sorry for noise.

In the future, would be great to have support for linux as documentation build target (as the latest toolchain there supports new Swift features quicker than macOS, e.g. Instant/Clock/Duration in Swift 5.7).

**Reply by finestructure** [reply]
We could actually trial this. I believe it's a small change to enable building docs on Linux if requested via the SPI manifest, we've just not had a use case yet. Caveat: I haven't actually tried building _any_ docs yet on Linux, so I'm not sure what state it's in. Have you?

**Reply by hassila** [reply]
Generation seems to work fine AFAICT by just checking the artifacts, I can test some more (just had to reinstall my local vm as it got corrupt....) if you like, but would be happy to be a guinea pig.

**Comment by hassila** [comment]
I've put up a .spi.yml for both `ordo-one/package-benchmark` and `ordo-one/package-frostflake` that should try it out:

E.g.:
```
version: 1
builder:
  configs:
  - documentation_targets: [Frostflake]
    platform: linux
```
